engine/src/main/java/org/terasology/rendering/nui/editor/NUIEditorContextMenuBuilder.java

```diff
@@ -0,0 +1,350 @@
+                        }, node, true);
+                    }
+                }
+            }
+        } else {
+            logger.warn("Unknown widget type: {}", type);
+        }
+    }
+
+    private void populateLayoutContextMenu(JsonTree node, ContextMenuLevel addLevel) {
+        String type = "";
+        JsonTree currentNode = node;
+        ClassMetadata layoutMetadata = null;
+        Class layoutHintType = null;
+
+        // Iterate down the tree and attempt to find a UILayout node.
+        while (layoutHintType == null) {
+            if (currentNode.hasSiblingWithKey("type")) {
+                type = (String) currentNode.getSiblingWithKey("type").getValue().getValue();
+                layoutMetadata = nuiManager.getWidgetMetadataLibrary().resolve(type, ModuleContext.getContext());
+
+                // If one is found, attempt to get the layout hint type.
+                try {
+                    layoutHintType = (Class) ReflectionUtil
+                        .getTypeParameter(layoutMetadata.getType().getGenericSuperclass(), 0);
+                } catch (NullPointerException e) {
+                    layoutHintType = null;
+                }
+            }
+            if (layoutHintType == null) {
+                if (currentNode.isRoot()) {
+                    break;
+                }
+                currentNode = (JsonTree) currentNode.getParent();
+            }
+        }
+
+        if (layoutHintType != null) {
+            for (Field field : layoutHintType.getDeclaredFields()) {
+                field.setAccessible(true);
+
+                // Add options for all @LayoutConfig-annotated fields.
+                if (field.isAnnotationPresent(LayoutConfig.class)) {
+                    String name = getNodeName(field);
+                    ClassMetadata finalMetadata = layoutMetadata;
+                    if (!node.hasChildWithKey(name)) {
+                        addLevel.addOption(name, n -> {
+                            try {
+                                createChild(name, node, field, finalMetadata);
+                            } catch (IllegalAccessException e) {
+                                return;
+                            }
+                            addContextMenuListeners.forEach(UpdateListener::onAction);
+                        }, node, true);
+                    }
+                }
+            }
+        } else {
+            logger.warn("Unknown layout type: {}", type);
+        }
+    }
+
+    private void populateGenericContextMenu(JsonTree node, ContextMenuLevel addLevel) {
+        if (node.hasSiblingWithKey("type")) {
+            // Get the ClassMetadata of the widget with the given type.
+            String type = (String) node.getSiblingWithKey("type").getValue().getValue();
+            ClassMetadata elementMetadata = nuiManager.getWidgetMetadataLibrary().resolve(type, ModuleContext.getContext());
```

Comment:
Same gets done in the method populateLayoutContextMenu above. Could be extracted into a method.